crates/sui-replay/src/batch_replay.rs

```diff
@@ -0,0 +1,155 @@
+        let cancel = cancel.clone();
+        tasks.push(run_task(
+            provider,
+            rpc_url_ref,
+            expensive_safety_check_config,
+            use_authority,
+            terminate_early,
+            cancel,
+        ));
+    }
+    let all_failed_transactions: Vec<_> = join_all(tasks).await.into_iter().flatten().collect();
+    info!(
+        "Finished replaying {} transactions, took {:?}",
+        provider.get_executed_count(),
+        cur_time.elapsed()
+    );
+    if all_failed_transactions.is_empty() {
+        info!("All transactions passed");
+    } else {
+        error!("Some transactions failed: {:?}", all_failed_transactions);
+    }
+}
+
+struct TransactionDigestProvider {
+    digests: Mutex<VecDeque<TransactionDigest>>,
+    total_count: usize,
+    executed_count: AtomicUsize,
+}
+
+impl TransactionDigestProvider {
+    pub fn new(digests: impl Iterator<Item = TransactionDigest>) -> Self {
+        let digests: VecDeque<_> = digests.collect();
+        let total_count = digests.len();
+        Self {
+            digests: Mutex::new(digests),
+            total_count,
+            executed_count: AtomicUsize::new(0),
+        }
+    }
+
+    pub fn get_total_count(&self) -> usize {
+        self.total_count
+    }
+
+    pub fn get_executed_count(&self) -> usize {
+        self.executed_count
+            .load(std::sync::atomic::Ordering::Relaxed)
+    }
+
+    /// Returns the index and digest of the next transaction, if any.
+    pub fn next_digest(&self) -> Option<(usize, TransactionDigest)> {
+        let next_digest = self.digests.lock().pop_front();
+        next_digest.map(|digest| {
+            let executed_count = self
+                .executed_count
+                .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+            (executed_count + 1, digest)
+        })
+    }
+}
+
+async fn run_task(
+    tx_digest_provider: Arc<TransactionDigestProvider>,
+    http_url: &str,
+    expensive_safety_check_config: ExpensiveSafetyCheckConfig,
+    use_authority: bool,
+    terminate_early: bool,
+    cancel: tokio_util::sync::CancellationToken,
+) -> Vec<ReplayEngineError> {
+    let total_count = tx_digest_provider.get_total_count();
+    let mut failed_transactions = vec![];
+    let mut executor = LocalExec::new_from_fn_url(http_url).await.unwrap();
+    while let Some((index, digest)) = tx_digest_provider.next_digest() {
+        if cancel.is_cancelled() {
+            break;
+        }
+        info!(
+            "[{}/{}] Replaying transaction {:?}...",
+            index, total_count, digest
+        );
+        executor = loop {
+            match executor.clone().reset_for_new_execution_with_client().await {
+                Ok(executor) => break executor,
+                Err(err) => {
+                    error!("Failed to reset executor: {:?}. Retrying in 3s", err);
+                    tokio::time::sleep(std::time::Duration::from_secs(3)).await;
```

Comment:
This is not going to be critical, but a benefit of using the cancellation token is that it offers a future that triggers on cancel, that you can `select!` against. At the moment, a given task can wait arbitrarily long because it failed to reset the executor or failed to execute the transaction, but you could build a helper to allow you to retry with cancellation on both of those tasks, so the job will always exit promptly.

I played around with something similar in this playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e9ba46c467db564693a95cb3cb7ee781